agent/Equalizer.py

```diff
@@ -0,0 +1,269 @@
+        # our player is leading, choosing a random card for now
+        if self.is_lead(partial_state):
+
+            # focus initially on voiding diamonds when leading
+            choice = self.void_out_suits(partial_state)
+
+        else:
+            # we're following, not leading
+            # so the idea here is to follow with our highest card for this suit (unless we're Hearts/Spades
+            choice = self.following_suit(partial_state)
+        return choice
+
+    def void_out_suits(self,
+                       partial_state: HeartsState):
+        """ This function selects the suit we want to try and void, the idea is to start with voiding either Clubs
+        or Diamonds (whichever is shortest) followed by Hearts and then Spades."""
+        # start by sorting out all of the clubs
+        suits = self.sort_suits(partial_state)
+        # the suit that we wish to pick a card from
+        suit_to_void = []
+
+        # todo right now leading is for when hearts has not been broken. may want to change strategy in future
+        # note that this nested if-else block can handle both cases of whether the agent is:
+        # - void in the suit the trick leader has
+        # - able to follow suit
+        if len(suits[1]) != 0 or len(suits[0]) != 0:
+            # neither diamonds or clubs have been voided, lead from one of these
+            if len(suits[1]) == 0 or len(suits[1]) > len(suits[0]) != 0:
+                # we want to void clubs first
+                suit_to_void = suits[0]
+            else:
+                # we want to void diamonds first
+                suit_to_void = suits[1]
+        else:
+            # clubs and diamonds are both void, lead with spades or hearts
+            if len(suits[3]) != 0:
+                # leading with hearts if we can
+                suit_to_void = suits[3]
+            else:
+                # no choice but to lead with spades
+                suit_to_void = suits[2]
+
+        # we now have our suit to pick a card from, want to lead with a low card
+        choice = self.get_low_card(suit_to_void)
+        return choice
+
+    def following_suit(self,
+                       partial_state: HeartsState):
+        """ This function will still attempt to void Clubs or Diamonds if it can, it may also try and play
+        the highest card in its suit, if it thinks it can afford to take on potential points."""
+        # depending on the suit we want to follow in different ways
+        suits = self.sort_suits(partial_state)
+
+        # so my idea is that if the suit we're following with is Hearts or Spades
+        # we will want to go with our lowest card
+        if len(suits[2]) != 0 or len(suits[3]) != 0:
+            # if we're void in the suit we're trying to follow, want to avoid playing spades
+            # going to play and potentially break hearts if we can
+            if len(suits[3]) != 0:
+                if len(suits[3]) == 0:
```

Comment:
Is this if check necessary? Checking if suits[3] (Hearts) is void is already done directly above.

Reply:
The if-statement above checks to see if either Hearts or Spades exists.  It's possible that only one of them exists, but once we get inside of that if-statement we don't actually know which one it is, hence the extra checking.

Reply:
I think i know what you are saying about not knowing which suit because of the 'or' in the if statement. I'm referring these nested if statements: 
`if len(suits[3]) != 0:` (line 142) -- Checks if suit 3 is not void
`if len(suits[3]) == 0:` (line 143) -- How can this ever be true if suit 3 is not void.
`print("hi-follow")`

Reply:
Oh I see.  That `if len(suits[3]) == 0:` is just some left over debug stuff that I thought I removed but apparently not.  My answer was referring to the if-statement above `if len(suits[3]) != 0:`, which I see now you weren't talking about.

Reply:
Removed the useless debug statements.